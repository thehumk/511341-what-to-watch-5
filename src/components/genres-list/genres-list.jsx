import React from 'react';
import PropTypes from 'prop-types';
import FilmsList from '../films-list/films-list';
import {propsForFilms} from '../../utils/prop-types';
import {connect} from 'react-redux';
import {ActionCreator} from '../../store/action';
import {QUANTITY_RENDER_FILMS, MAX_QUANTITY_GENRES_TO_RENDER} from '../../utils/const';
import {getFilteredFilms} from '../../store/selectors';

const GenresList = (props) => {
  const {filteredFilms, activeGenre, renderFilmsCount, genresList, changeActiveGenre, changeRenderFilmsCount} = props;

  return (
    <>
      <ul className="catalog__genres-list">
        {genresList.slice(0, MAX_QUANTITY_GENRES_TO_RENDER).map((elem, i) => (
          <li key={i} className={`catalog__genres-item ${activeGenre === elem ? `catalog__genres-item--active` : ``}`} onClick={() => {
            changeActiveGenre(elem);
            changeRenderFilmsCount(QUANTITY_RENDER_FILMS);
          }}>
            <a className="catalog__genres-link">{elem}</a>
          </li>
        ))}
      </ul>

      <FilmsList
        films={filteredFilms}
        renderFilmsCount={renderFilmsCount}
        changeRenderFilmsCount={changeRenderFilmsCount}
      />
    </>
  );
};

GenresList.propTypes = {
  filteredFilms: PropTypes.arrayOf(propsForFilms).isRequired,
  activeGenre: PropTypes.string.isRequired,
  renderFilmsCount: PropTypes.number.isRequired,
  genresList: PropTypes.array.isRequired,
  changeActiveGenre: PropTypes.func.isRequired,
  changeRenderFilmsCount: PropTypes.func.isRequired,
};

const mapStateToProps = ({DATA, FILMS_STATUS}) => ({
  activeGenre: FILMS_STATUS.activeGenre,
  filteredFilms: getFilteredFilms({films: DATA.films, genre: FILMS_STATUS.activeGenre}),
  renderFilmsCount: FILMS_STATUS.renderFilmsCount,
  genresList: DATA.genresList,
});

const mapDispatchToProps = (dispatch) => ({
  changeActiveGenre(genre) {
    dispatch(ActionCreator.changeActiveGenre(genre));
  },
  changeRenderFilmsCount(count) {
    dispatch(ActionCreator.changeRenderFilmsCount(count));
  },
});

export {GenresList};
export default connect(mapStateToProps, mapDispatchToProps)(GenresList);
