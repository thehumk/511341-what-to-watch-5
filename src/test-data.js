export const noop = () => {};

export const MATCH = {params: {id: `1`}};

export const mockComments = [
  {
    comment: `123`,
    date: `2020-10-29T13:38:44.769Z`,
    id: 1,
    rating: 4.6,
    user: {
      id: 12,
      name: `Isaac`
    },
  },
  {
    comment: `1234`,
    date: `2020-10-29T13:38:45.769Z`,
    id: 2,
    rating: 4.5,
    user: {
      id: 13,
      name: `Isa`
    },
  }
];

/* eslint-disable */

export const mockFilms = [
  {
    background_color: `#73B39A`,
    background_image: `https://assets.htmlacademy.ru/intensives/javascript-3/film/background/bronson.jpg`,
    description: `A young man who was sentenced to seven years in prison for robbing a post office ends up spending three decades in solitary confinement. During this time, his own personality is supplanted by his alter-ego, Charles Bronson.`,
    director: `Nicolas Winding Refn`,
    genre: `Action`,
    id: 1,
    is_favorite: true,
    name: `Bronson`,
    poster_image: `https://assets.htmlacademy.ru/intensives/javascript-3/film/poster/bronson.jpg`,
    preview_image: `https://assets.htmlacademy.ru/intensives/javascript-3/film/preview/bronson.jpg`,
    preview_video_link: `https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm`,
    rating: 3.6,
    released: 2008,
    run_time: 92,
    scores_count: 109661,
    starring: [`Tom Hardy`, `Kelly Adams`, `Luing Andrews`],
    video_link: `http://media.xiph.org/mango/tears_of_steel_1080p.webm`,
  },
  {
    background_color: `#73B39A`,
    background_image: `https://assets.htmlacademy.ru/intensives/javascript-3/film/background/bronson.jpg`,
    description: `A young man who was sentenced to seven years in prison for robbing a post office ends up spending three decades in solitary confinement. During this time, his own personality is supplanted by his alter-ego, Charles Bronson.`,
    director: `Nicolas Winding Refn`,
    genre: `Drama`,
    id: 2,
    is_favorite: false,
    name: `Bronson`,
    poster_image: `https://assets.htmlacademy.ru/intensives/javascript-3/film/poster/bronson.jpg`,
    preview_image: `https://assets.htmlacademy.ru/intensives/javascript-3/film/preview/bronson.jpg`,
    preview_video_link: `https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm`,
    rating: 3.6,
    released: 2008,
    run_time: 92,
    scores_count: 109661,
    starring: [`Tom Hardy`, `Kelly Adams`, `Luing Andrews`],
    video_link: `http://media.xiph.org/mango/tears_of_steel_1080p.webm`,
  },
];

/* eslint-enable */

export const state = {
  DATA: {
    films: mockFilms,
    promoFilm: mockFilms[0],
    genresList: [`All`, `Action`, `Drama`],
    isApplicationReady: false,
  },
  FILMS_STATUS: {
    activeGenre: `All`,
    renderFilmsCount: 8,
  },
  USER: {
    authorizationStatus: `NO_AUTH`,
    user: null,
  },
  COMMENTS: {
    comments: mockComments,
  },
};
