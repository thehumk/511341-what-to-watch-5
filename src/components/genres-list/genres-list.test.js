import React from 'react';
import {BrowserRouter} from 'react-router-dom';
import renderer from 'react-test-renderer';
import {GenresList} from './genres-list';
import {noop, mockFilms, state} from '../../test-data';

it(`Should GenresList render correctly`, () => {
  const tree = renderer.create(
      <BrowserRouter>
        <GenresList
          filteredFilms={mockFilms}
          activeGenre={state.FILMS_STATUS.activeGenre}
          renderFilmsCount={state.FILMS_STATUS.renderFilmsCount}
          genresList={state.DATA.genresList}
          changeActiveGenre={noop}
          changeRenderFilmsCount={noop}
        />
      </BrowserRouter>
  ).toJSON();

  expect(tree).toMatchSnapshot();
});
