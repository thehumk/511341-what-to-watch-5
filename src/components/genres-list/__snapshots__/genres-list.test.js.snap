// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should GenresList render correctly 1`] = `
Array [
  <ul
    className="catalog__genres-list"
  >
    <li
      className="catalog__genres-item catalog__genres-item--active"
      onClick={[Function]}
    >
      <a
        className="catalog__genres-link"
      >
        All
      </a>
    </li>
    <li
      className="catalog__genres-item "
      onClick={[Function]}
    >
      <a
        className="catalog__genres-link"
      >
        Action
      </a>
    </li>
    <li
      className="catalog__genres-item "
      onClick={[Function]}
    >
      <a
        className="catalog__genres-link"
      >
        Drama
      </a>
    </li>
  </ul>,
  <div
    className="catalog__movies-list"
  >
    <article
      className="small-movie-card catalog__movies-card"
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
    >
      <a
        className="small-movie-card__link"
        href="/films/1"
        onClick={[Function]}
      >
        <div
          className="small-movie-card__image"
        >
          <video
            height="175"
            muted={true}
            poster="https://assets.htmlacademy.ru/intensives/javascript-3/film/preview/bronson.jpg"
            src="https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm"
            width="280"
          />
        </div>
        <h3
          className="small-movie-card__title"
        >
          <span>
            Bronson
          </span>
        </h3>
      </a>
    </article>
    <article
      className="small-movie-card catalog__movies-card"
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
    >
      <a
        className="small-movie-card__link"
        href="/films/2"
        onClick={[Function]}
      >
        <div
          className="small-movie-card__image"
        >
          <video
            height="175"
            muted={true}
            poster="https://assets.htmlacademy.ru/intensives/javascript-3/film/preview/bronson.jpg"
            src="https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm"
            width="280"
          />
        </div>
        <h3
          className="small-movie-card__title"
        >
          <span>
            Bronson
          </span>
        </h3>
      </a>
    </article>
  </div>,
  <div
    className="catalog__more"
  >
    
  </div>,
]
`;
